#!/bin/bash
i=$1
/fml/chones/local/bin/bcftools view $i -h | head -2 > ${i/.vcf.gz/.header};
cat ../contigs.helera1_demo.header >> ${i/.vcf.gz/.header};
/fml/chones/local/bin/bcftools view $i -h | awk 'NR > 2' >> ${i/.vcf.gz/.header};
awk '/^#/;/"All filters passed">/ {} ; /reference and alternate alleles/ {print "##INFO=<ID=INDEL,Number=0,Type=Flag,Description=\"\">"};/Dosage/ {print "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">";print "##FORMAT=<ID=PG,Number=1,Type=String,Description=\"Best Guessed Genotype\">";}' ${i/.vcf.gz/.header}  >  ${i/.vcf.gz/.header}.1; mv ${i/.vcf.gz/.header}.1 ${i/.vcf.gz/.header}
#cp ${i/.vcf/.header} ${i/.vcf/.PL.vcf}; /fml/chones/local/bin/bcftools view $i -H | awk 'BEGIN {OFS="\t"};{$9="GT:GP:DS:PL"; for (i=10; i<=NF; i++) {split($i,field,":");split(field[2],gp,",");pl1=-log(gp[1])/log(10)*10;pl2=-log(gp[2])/log(10)*10;pl3=-log(gp[3])/log(10)*10;if(pl1=="inf") {pl1=255}; if(pl2=="inf") {pl2=255};if(pl3=="inf") {pl3=255};$i=$i":"int(pl1+0.5)","int(pl2+0.5)","int(pl3+0.5)};print $0}' >> ${i/.vcf/.PL.vcf}; rm ${i/.vcf/.PL.vcf.gz}; bgzip -@ 32 ${i/.vcf/.PL.vcf}; tabix -f ${i/.vcf/.PL.vcf.gz};
cp ${i/.vcf.gz/.header} ${i/.vcf.gz/.PL.vcf}; /fml/chones/local/bin/bcftools view $i -H | awk 'BEGIN {OFS="\t"};{$9="GT:PG:GP:DS:PL"; for (i=10; i<=NF; i++)  {split($i,field,":");split(field[2],gp,",");ngt=field[1];if(gp[1]>gp[2] && gp[1]>gp[3]) {ngt="0/0"}; if(gp[2]>gp[1] && gp[2]>gp[3]) {ngt="0/1"}; if (gp[3]>gp[1] && gp[3] > gp[2]) {ngt="1/1"}; pl1=-log(gp[1])/log(10)*10;pl2=-log(gp[2])/log(10)*10;pl3=-log(gp[3])/log(10)*10;if(pl1=="inf") {pl1=255}; if(pl2=="inf") {pl2=255};if(pl3=="inf") {pl3=255};$i=ngt":"$i":"int(pl1+0.5)","int(pl2+0.5)","int(pl3+0.5)};print $0}' >> ${i/.vcf.gz/.PL.vcf}; rm ${i/.vcf.gz/.PL.vcf.gz}; bgzip -@ 32 ${i/.vcf.gz/.PL.vcf}; tabix -f ${i/.vcf.gz/.PL.vcf.gz};
